Userland/Services/FileSystemAccessServer/ClientConnection.cpp

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2021, timmot <tiwwot@protonmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "DownloadWidget.h"
+#include <AK/Debug.h>
+#include <FileSystemAccessServer/ClientConnection.h>
+#include <LibC/fcntl.h>
+#include <LibCore/File.h>
+#include <LibCore/StandardPaths.h>
+#include <LibGUI/Application.h>
+#include <LibGUI/FilePicker.h>
+#include <LibGUI/MessageBox.h>
+
+namespace FileSystemAccessServer {
+
+static HashMap<int, RefPtr<ClientConnection>> s_connections;
+
+ClientConnection::ClientConnection(NonnullRefPtr<Core::LocalSocket> socket, int client_id)
+    : IPC::ClientConnection<FileSystemAccessClientEndpoint, FileSystemAccessServerEndpoint>(*this, move(socket), client_id)
+{
+    s_connections.set(client_id, *this);
+}
+
+ClientConnection::~ClientConnection()
+{
+}
+
+void ClientConnection::die()
+{
+    s_connections.remove(client_id());
+    exit(0);
+}
+
+Messages::FileSystemAccessServer::RequestFileReadResponse ClientConnection::request_file_read(String const& path)
+{
+    // TODO: I would like to automatically include a pointer to the caller object so I can call Object::class_name
+    String sender_name = "Browser";
+
+    auto app = GUI::Application::construct(0, nullptr);
+    auto result = GUI::MessageBox::show(nullptr, String::formatted("Give {} access to read from \"{}\"?", sender_name, path), "File Read Requested", GUI::MessageBox::Type::Warning, GUI::MessageBox::InputType::YesNo);
+    app->exec();
+    app->quit();
+
+    if (result == GUI::MessageBox::ExecYes) {
+        auto fd = open(path.characters(), O_RDONLY);
+
+        // TODO: perform error checking
+        // TODO: if root file...
+
+        return { -1, IPC::File(fd, IPC::File::CloseAfterSending) };
```

Comment:
```suggestion
        return Optional<IPC::File> { IPC::File(fd, IPC::File::CloseAfterSending) };
```

Reply:
Thank you Gunnar!

I was having trouble compiling IPC::File by itself, turns out I was using Optional incorrectly.